### PR TITLE
Fix `Index.difference` to match with pandas

### LINKED ISSUE #14013: [FEA]: Preserve RangeIndex in more setops cases
**Is your feature request related to a problem? Please describe.**
Sometimes the result of `difference`, `intersection`, `union` could still be represented as a `RangeIndex

```python
In [3]: from cudf import *

In [4]:         idx = Index(range(4))[::-1]
   ...:         other = Index(range(3, 4))
   ...: 
   ...:         result = idx.difference(other)

In [5]: idx
Out[5]: RangeIndex(start=3, stop=-1, step=-1)

In [6]: other
Out[6]: RangeIndex(start=3, stop=4, step=1)

In [7]: result
Out[7]: Int64Index([0, 1, 2], dtype='int64')
```

**Describe the solution you'd like**
In the case above, `RangeIndex(0, 3, 1)` could have been returned

**Describe alternatives you've considered**
Casting directly to a RangeIndex

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### LINKED ISSUE #14019: [BUG] Incorrect `name` selection in `Index.difference` when an object with no-name is passed
**Describe the bug**
When a difference is calculated with an object that has no name, the result should have the name of the Index object itself.

**Steps/Code to reproduce bug**
```python
In [1]: import cudf

In [2]: s = cudf.Index(['a', 'b', 'c'], name="abc")

In [3]: s.difference([])
Out[3]: StringIndex(['a' 'b' 'c'], dtype='object')

In [4]: s.to_pandas().difference([])
Out[4]: Index(['a', 'b', 'c'], dtype='object', name='abc')
```

**Expected behavior**
```
In [4]: s.difference([])
Out[4]: Index(['a', 'b', 'c'], dtype='object', name='abc')
```

**Environment overview (please complete the following information)**
 - Environment location: [Bare-metal]
 - Method of cuDF install: [from source]

### LINKED ISSUE #14022: [BUG] `CategoricalIndex` leading to incorrect results in `Index.difference`
**Describe the bug**
When a `CategoricalIndex` is passed as an input to `Index.difference`, the values aren't being compared as a decategorized column.

**Steps/Code to reproduce bug**
```python
In [1]: import cudf

In [2]: s = cudf.Index(['a', 'b', 'c'])

In [3]: c = s.astype('category')

In [4]: s.difference(s)
Out[4]: StringIndex([], dtype='object')

In [5]: s.difference(c)
Out[5]: StringIndex(['a' 'b' 'c'], dtype='object')


In [6]: s.to_pandas().difference(c.to_pandas())
Out[6]: Index([], dtype='object')

```
**Expected behavior**
```python

In [6]: s.difference(c)
Out[6]: Index([], dtype='object')
```

**Environment overview (please complete the following information)**
 - Environment location: [Bare-metal]
 - Method of cuDF install: [from source]

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description

This PR fixes `Index.difference` in following ways:

- [x] Fixes `name` preservation by correctly evaluating the name of two input objects, closes #14019
- [x] Fixes `is_mixed_with_object_dtype` handling that will resolve incorrect results for `CategoricalIndex`, closes #14022
- [x] Raises errors for invalid input types, the error messages are an exact match to pandas error messages for parity.
- [x] Introduce a `Range._try_reconstruct_range_index` that will try to re-construct a `RangeIndex` out of an `Int..Index` to save memory- this is on parity with pandas. closes #14013 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -724,6 +724,29 @@ def _intersection(self, other, sort=False):
...
+    def difference(self, other, sort=None):
+        if isinstance(other, RangeIndex) and self.equals(other):
+            return self[:0]._get_reconciled_name_object(other)
+
+        return self._try_reconstruct_range_index(
+            super().difference(other, sort=sort)
+        )
+
+    def _try_reconstruct_range_index(self, index):
+        if index.dtype.kind == "f":
```
Comment (wence-) [inline python/cudf/cudf/core/index.py:737]: Under some circumstances, `super().difference(...)` can already return a `RangeIndex` (e.g. `RangeIndex.difference(StringIndex)`)

File: python/cudf/cudf/core/index.py [hunk 2]
```diff
@@ -724,6 +724,29 @@ def _intersection(self, other, sort=False):
...
+            return index
+        # Evenly spaced values can return a
+        # RangeIndex instead of Int64Index
+        unique_diffs = index.to_frame(name="None").diff()["None"].unique()
```
Comment (wence-) [inline python/cudf/cudf/core/index.py:741]: Prefer cupy?
```
In [1]: import rmm

In [2]: rmm.reinitialize(pool_allocator=True)

In [3]: import cudf
c = 
In [4]: c = cudf.RangeIndex(0, 10_000_000, 2)._as_int_index()
im
In [5]: import cupy

In [6]: %timeit cupy.unique(cupy.diff(c.values));
2.72 ms ± 3.61 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: %timeit c.to_frame(name="_").diff()["_"].unique();
4.97 ms ± 3.62 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: c = cudf.RangeIndex(0, 100_000_000, 3)._as_int_index()

In [9]: %timeit cupy.unique(cupy.diff(c.values));
15 ms ± 8.33 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: %timeit c.to_frame(name="_").diff()["_"].unique();
27.3 ms ± 34.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [11]: c = cudf.RangeIndex(0, 1000_000, 3)._as_int_index()

In [12]: %timeit cupy.unique(cupy.diff(c.values));
740 µs ± 1.31 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [13]: %timeit c.to_frame(name="_").diff()["_"].unique();
873 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Review by wence- — CHANGES_REQUESTED** [review]
Looks good overall, but I think there's a missing corner-case in the reconstruct logic, and cupy appears to be a bit fast than conversion to a frame for the `unique` call

**Comment by mroeschke** [comment]
Does this close https://github.com/rapidsai/cudf/issues/14013?

**Comment by galipremsagar** [comment]
> Does this close #14013?

Yup. Linked to the PR.

File: python/cudf/cudf/core/index.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Good analysis here, switched to cupy 👍

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Added additional checks

**Review by wence- — APPROVED** [review]
Thanks! (Didn't know an index could have nulls!)

**Pull request merged**